staging/src/k8s.io/apiserver/pkg/util/peerproxy/exclusion_filter.go

```diff
@@ -0,0 +1,370 @@
+	for existingGVR, v := range entry.GVRs {
+		gv := schema.GroupVersion{Group: existingGVR.Group, Version: existingGVR.Version}
+		if _, skip := excluded[gv]; skip {
+			peerDiscoveryChanged = true
+			continue
+		}
+		filteredGVRs[existingGVR] = v
+	}
+
+	if !peerDiscoveryChanged {
+		return entry, false
+	}
+
+	// Filter the group discovery list
+	filteredGroups := h.filterGroupDiscovery(entry.GroupDiscovery, excluded)
+	return PeerDiscoveryCacheEntry{
+		GVRs:           filteredGVRs,
+		GroupDiscovery: filteredGroups,
+	}, true
+}
+
+func (h *peerProxyHandler) filterGroupDiscovery(groupDiscoveries []apidiscoveryv2.APIGroupDiscovery, excluded map[schema.GroupVersion]struct{}) []apidiscoveryv2.APIGroupDiscovery {
+	var filteredDiscovery []apidiscoveryv2.APIGroupDiscovery
+	for _, groupDiscovery := range groupDiscoveries {
+		filteredGroup := apidiscoveryv2.APIGroupDiscovery{
+			ObjectMeta: groupDiscovery.ObjectMeta,
+		}
+
+		for _, version := range groupDiscovery.Versions {
+			gv := schema.GroupVersion{Group: groupDiscovery.Name, Version: version.Version}
+			if _, found := excluded[gv]; found {
+				// This version is excluded, skip it
+				continue
+			}
+			filteredGroup.Versions = append(filteredGroup.Versions, version)
+		}
+
+		// Only add the group to the final list if it still has any versions left
+		if len(filteredGroup.Versions) > 0 {
+			filteredDiscovery = append(filteredDiscovery, filteredGroup)
+		}
+	}
+	return filteredDiscovery
+}
+
+// RunGVDeletionWorkers starts worker goroutines to process the GV deletion queue.
+func (h *peerProxyHandler) RunGVDeletionWorkers(ctx context.Context, workers int) {
+	defer h.gvDeletionQueue.ShutDown()
+
+	klog.Infof("Starting %d GV deletion worker(s)", workers)
+	for i := 0; i < workers; i++ {
+		go func(workerID int) {
+			for h.processNextGVDeletion() {
+				select {
+				case <-ctx.Done():
+					klog.Infof("GV deletion worker %d shutting down", workerID)
+					return
+				default:
+				}
+			}
+		}(i)
+	}
+
+	<-ctx.Done()
+	klog.Info("Shutting down GV deletion workers")
+}
+
+// processNextGVDeletion processes a single item from the GV deletion queue.
+func (h *peerProxyHandler) processNextGVDeletion() bool {
+	gvString, shutdown := h.gvDeletionQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer h.gvDeletionQueue.Done(gvString)
+
+	gv, err := schema.ParseGroupVersion(gvString)
+	if err != nil {
+		klog.Errorf("Failed to parse GroupVersion %q: %v", gvString, err)
+		return true
+	}
+
+	h.markGVForDeletionIfUnused(gv)
+	return true
+}
+
+// markGVForDeletionIfUnused checks if a group-version is still in use.
+// If not, it marks it for deletion by setting a timestamp.
+func (h *peerProxyHandler) markGVForDeletionIfUnused(gv schema.GroupVersion) {
+	// Check if GV is still in use BEFORE acquiring the lock to avoid holding
+	// the lock during the potentially slow iteration over informer stores.
+	if h.isGVUsed(gv) {
+		return
+	}
+
+	h.excludedGVsMu.Lock()
+	defer h.excludedGVsMu.Unlock()
```

Comment:
Do we need to lock the call to `isGVUsed` ?  If so can we document that isDVUsed is unsafe?  If not, can we narrow down what code we hold the lock for?

Reply:
We dont need to lock for `isGVUsed` since it only invoked a list() on informers. But since acquiring the lock (which happens after the first call to `isGVUsed`) may take some time, we invoke`isGVUsed` once again to be sure we are looking at the latest state, to avoid any race. 
If instead, we first acquire the lock and directly call `isGVUsed()` once after that - that may cause the lock to be help up for a long time since `isGVUsed()` does a list() on both crd and apiservice informer.

Reply:
Maybe that race is super unlikely so it maybe ok to just not do the `isGVUsed(gv)` a second time? Happy to remove it if you think its an overkill